File: src/services/regulations.ts
import type { ReviewableField, ReviewSource, EvidenceEntry } from '../types';
import { get, post } from './api';
import { reviewSources } from '../data/regulations';

// Review sources catalogue (static — used for source selector UI and prev/next)
export function getReviewSources(): ReviewSource[] {
  return reviewSources;
}

export function getReviewSourceById(id: number): ReviewSource | undefined {
  return reviewSources.find(s => s.id === id);
}

// ---------------------------------------------------------------------------
// Regulation detail from API (includes fields, source text, evidence)
// ---------------------------------------------------------------------------

export interface RegulationResponse {
  id: number;
  sourceId: number;
  title: string;
  regulatoryBody: string;
  jurisdiction: string;
  topic: string;
  summary: string | null;
  status: string;
  fields: ReviewableField[];
  sourceText: string | null;
  sourceName: string | null;
  flag: string | null;
  docType: string | null;
  country: string | null;
}

export async function getRegulationById(regId: number): Promise<RegulationResponse | null> {
  try {
    const res = await get<{ data: RegulationResponse }>(`/api/regulations/${regId}`);
    return res.data;
  } catch { return null; }
}

// Compatibility: get fields for a source (finds regulation by source_id)
export async function getFieldsForSource(sourceId: number): Promise<ReviewableField[]> {
  // First get the regulation(s) for this source
  const regsRes = await get<{ data: { id: number }[] }>(`/api/sources/${sourceId}/regulations`);
  if (!regsRes.data.length) return [];
  // Get fields from the first regulation
  const regId = regsRes.data[0].id;
  const regRes = await get<{ data: { fields: ReviewableField[] } }>(`/api/regulations/${regId}`);
  return regRes.data.fields;
}

// Evidence from API
export async function getEvidenceForField(fieldId: number): Promise<{
  excerpt: string; section: string | null; sourceReference: string | null; immutable: boolean;
} | null> {
  try {
    const res = await get<{ data: { excerpt: string; section: string | null; sourceReference: string | null; immutable: boolean } }>(`/api/evidence/${fieldId}`);
    return res.data;
  } catch { return null; }
}

// Source text from API
export async function getSourceTextForSource(sourceId: number): Promise<string | null> {
  try {
    const res = await get<{ data: { sourceText: string | null } }>(`/api/sources/${sourceId}`);
    return res.data.sourceText;
  } catch { return null; }
}

// Keep static evidence map for highlight ranges (source-text specific, not from DB)
export function getHighlightRanges(): Record<string, [number, number]> {
  return {
    Title: [2, 4], Summary: [11, 15], 'Products Impacted': [18, 20],
    'Sector Impact': [41, 42], Likelihood: [45, 46], 'Comment Deadline': [49, 50],
    'Effective Date': [45, 46],
  };
}

// ---------------------------------------------------------------------------
// Field-level review actions (now via /api/fields endpoints)
// ---------------------------------------------------------------------------

export async function acceptField(_sourceId: number, fieldId: number): Promise<ReviewableField> {
  const res = await post<{ data: ReviewableField }>(`/api/fields/${fieldId}/review`, {
    fieldId, decision: 'Accepted',
  });
  return res.data;
}

export async function rejectField(_sourceId: number, fieldId: number): Promise<ReviewableField> {
  const res = await post<{ data: ReviewableField }>(`/api/fields/${fieldId}/review`, {
    fieldId, decision: 'Rejected',
  });
  return res.data;
}

export async function flagField(_sourceId: number, fieldId: number): Promise<ReviewableField> {
  const res = await post<{ data: ReviewableField }>(`/api/fields/${fieldId}/review`, {
    fieldId, decision: 'Flagged',
  });
  return res.data;
}

export async function resetField(_sourceId: number, fieldId: number): Promise<ReviewableField> {
  const res = await post<{ data: ReviewableField }>(`/api/fields/${fieldId}/review`, {
    fieldId, decision: 'Reset',
  });
  return res.data;
}

export async function editFieldValue(_sourceId: number, fieldId: number, newValue: string): Promise<ReviewableField> {
  const res = await post<{ data: ReviewableField }>(`/api/fields/${fieldId}/review`, {
    fieldId, decision: 'Edited', newValue,
  });
  return res.data;
}

export async function addFieldComment(_sourceId: number, fieldId: number, comment: string): Promise<ReviewableField> {
  const res = await post<{ data: ReviewableField }>(`/api/fields/${fieldId}/review`, {
    fieldId, decision: 'Comment', comment,
  });
  return res.data;
}

export async function acceptAllFields(sourceId: number): Promise<ReviewableField[]> {
  // Find the regulation for this source, then accept all
  const regsRes = await get<{ data: { id: number }[] }>(`/api/sources/${sourceId}/regulations`);
  if (!regsRes.data.length) return [];
  const regId = regsRes.data[0].id;
  const res = await post<{ data: ReviewableField[] }>(`/api/regulations/${regId}/review/accept-all`, {});
  return res.data;
}

export function getReviewStats(fields: ReviewableField[]) {
  return {
    total: fields.length,
    accepted: fields.filter(f => f.status === 'Accepted').length,
    rejected: fields.filter(f => f.status === 'Rejected').length,
    flagged: fields.filter(f => f.status === 'Flagged').length,
    pending: fields.filter(f => f.status === 'Pending').length,
  };
}

File: backend/routers/regulations.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from database import get_connection
from datetime import datetime

router = APIRouter()


def _now() -> str:
    d = datetime.now()
    return f"{d.strftime('%b')} {d.day}, 2026 {d.strftime('%H:%M')}"


def _field_dict(row) -> dict:
    return {
        "id": row["id"], "regulationId": row["regulation_id"], "sourceId": row["source_id"],
        "category": row["category"], "field": row["field_name"],
        "extractedValue": row["extracted_value"], "reviewedValue": row["reviewed_value"],
        "confidence": row["confidence"], "status": row["status"],
        "comment": row["comment"], "reviewedAt": row["reviewed_at"],
    }


def _reg_dict(row) -> dict:
    return {
        "id": row["id"], "sourceId": row["source_id"], "title": row["title"],
        "regulatoryBody": row["regulatory_body"], "jurisdiction": row["jurisdiction"],
        "topic": row["topic"], "summary": row["summary"], "status": row["status"],
        "createdAt": row["created_at"], "updatedAt": row["updated_at"],
    }


@router.get("/api/regulations")
def list_regulations(source_id: int | None = None):
    conn = get_connection()
    if source_id:
        rows = conn.execute("SELECT * FROM regulations WHERE source_id = ? ORDER BY id", (source_id,)).fetchall()
    else:
        rows = conn.execute("SELECT * FROM regulations ORDER BY id").fetchall()
    conn.close()
    return {"data": [_reg_dict(r) for r in rows]}


@router.get("/api/regulations/{reg_id}")
def get_regulation(reg_id: int):
    conn = get_connection()
    reg = conn.execute("SELECT * FROM regulations WHERE id = ?", (reg_id,)).fetchone()
    if not reg:
        conn.close(); raise HTTPException(404, f"Regulation {reg_id} not found")
    fields = conn.execute("SELECT * FROM regulation_fields WHERE regulation_id = ? ORDER BY id", (reg_id,)).fetchall()
    source = conn.execute("SELECT * FROM sources WHERE id = ?", (reg["source_id"],)).fetchone()
    # Attach evidence to each field
    field_list = []
    for f in fields:
        fd = _field_dict(f)
        ev = conn.execute("SELECT * FROM evidence WHERE field_id = ?", (f["id"],)).fetchone()
        if ev:
            fd["evidence"] = ev["excerpt"]
            fd["evidenceSection"] = ev["section"]
            fd["evidenceReference"] = ev["source_reference"]
        else:
            fd["evidence"] = ""
            fd["evidenceSection"] = None
            fd["evidenceReference"] = None
        field_list.append(fd)
    conn.close()
    d = _reg_dict(reg)
    d["fields"] = field_list
    d["sourceText"] = source["source_text"] if source else None
    d["sourceName"] = source["source_name"] if source else None
    d["flag"] = source["flag"] if source else None
    d["docType"] = source["doc_type"] if source else None
    d["country"] = source["country"] if source else None
    return {"data": d}


class RegPatch(BaseModel):
    status: str | None = None
    summary: str | None = None


@router.patch("/api/regulations/{reg_id}")
def patch_regulation(reg_id: int, body: RegPatch):
    conn = get_connection()
    reg = conn.execute("SELECT * FROM regulations WHERE id = ?", (reg_id,)).fetchone()
    if not reg:
        conn.close(); raise HTTPException(404)
    updates, params = [], []
    if body.status: updates.append("status = ?"); params.append(body.status)
    if body.summary: updates.append("summary = ?"); params.append(body.summary)
    if updates:
        updates.append("updated_at = ?"); params.append(_now()); params.append(reg_id)
        conn.execute(f"UPDATE regulations SET {', '.join(updates)} WHERE id = ?", params)
        conn.commit()
    updated = conn.execute("SELECT * FROM regulations WHERE id = ?", (reg_id,)).fetchone()
    conn.close()
    return {"data": _reg_dict(updated)}


# -- Field-level endpoints ---------------------------------------------------

@router.get("/api/fields/{field_id}")
def get_field(field_id: int):
    conn = get_connection()
    row = conn.execute("SELECT * FROM regulation_fields WHERE id = ?", (field_id,)).fetchone()
    if not row:
        conn.close(); raise HTTPException(404)
    ev = conn.execute("SELECT * FROM evidence WHERE field_id = ?", (field_id,)).fetchone()
    conn.close()
    d = _field_dict(row)
    if ev:
        d["evidence"] = {"id": ev["id"], "excerpt": ev["excerpt"], "section": ev["section"],
                         "sourceReference": ev["source_reference"], "immutable": bool(ev["immutable"])}
    return {"data": d}


class FieldPatch(BaseModel):
    status: str | None = None
    reviewedValue: str | None = None
    comment: str | None = None


@router.patch("/api/fields/{field_id}")
def patch_field(field_id: int, body: FieldPatch):
    conn = get_connection()
    row = conn.execute("SELECT * FROM regulation_fields WHERE id = ?", (field_id,)).fetchone()
    if not row:
        conn.close(); raise HTTPException(404)
    ts = _now()
    updates, params = [], []
    if body.status is not None:
        updates += ["status = ?", "reviewed_at = ?"]; params += [body.status, ts]
    if body.reviewedValue is not None:
        updates.append("reviewed_value = ?"); params.append(body.reviewedValue)
        if body.status is None:
            updates += ["status = ?", "reviewed_at = ?"]; params += ["Accepted", ts]
    if body.comment is not None:
        updates.append("comment = ?"); params.append(body.comment)
        if "reviewed_at = ?" not in updates:
            updates.append("reviewed_at = ?"); params.append(ts)
    if updates:
        decision = body.status or ("Edited" if body.reviewedValue else "Comment")
        conn.execute(
            "INSERT INTO review_decisions (field_id,decision,previous_value,new_value,comment,created_at) VALUES (?,?,?,?,?,?)",
            (field_id, decision, row["reviewed_value"] or row["extracted_value"], body.reviewedValue, body.comment, ts),
        )
        params.append(field_id)
        conn.execute(f"UPDATE regulation_fields SET {', '.join(updates)} WHERE id = ?", params)
        conn.commit()
    updated = conn.execute("SELECT * FROM regulation_fields WHERE id = ?", (field_id,)).fetchone()
    conn.close()
    return {"data": _field_dict(updated)}


class ReviewBody(BaseModel):
    fieldId: int
    decision: str
    newValue: str | None = None
    comment: str | None = None


@router.post("/api/fields/{field_id}/review")
def review_field(field_id: int, body: ReviewBody):
    conn = get_connection()
    row = conn.execute("SELECT * FROM regulation_fields WHERE id = ?", (field_id,)).fetchone()
    if not row:
        conn.close(); raise HTTPException(404, f"Field {field_id} not found")
    ts = _now()
    status_map = {"Accepted": "Accepted", "Rejected": "Rejected", "Flagged": "Flagged",
                  "Edited": "Accepted", "Reset": "Pending", "Comment": None}
    new_status = status_map.get(body.decision, body.decision)

    updates, params = [], []
    if new_status is not None:
        updates += ["status = ?", "reviewed_at = ?"]; params += [new_status, ts]
    if body.newValue is not None:
        updates.append("reviewed_value = ?"); params.append(body.newValue)
    if body.comment is not None:
        updates.append("comment = ?"); params.append(body.comment)
    if body.decision == "Reset":
        updates = ["status = 'Pending'", "reviewed_value = NULL", "comment = NULL", "reviewed_at = NULL"]
        params = []

    conn.execute(
        "INSERT INTO review_decisions (field_id,decision,previous_value,new_value,comment,created_at) VALUES (?,?,?,?,?,?)",
        (field_id, body.decision, row["reviewed_value"] or row["extracted_value"], body.newValue, body.comment, ts),
    )
    params.append(field_id)
    conn.execute(f"UPDATE regulation_fields SET {', '.join(updates)} WHERE id = ?", params)
    conn.commit()
    updated = conn.execute("SELECT * FROM regulation_fields WHERE id = ?", (field_id,)).fetchone()
    conn.close()
    return {"data": _field_dict(updated)}


# -- Compatibility routes (used by current frontend) -------------------------

@router.post("/api/regulations/{reg_id}/review")
def review_field_compat(reg_id: int, body: ReviewBody):
    """Compatibility: accepts reg_id in URL but routes to field-level review."""
    return review_field(body.fieldId, body)


@router.post("/api/regulations/{reg_id}/review/accept-all")
def accept_all_fields(reg_id: int):
    conn = get_connection()
    ts = _now()
    conn.execute(
        "UPDATE regulation_fields SET status = 'Accepted', reviewed_at = ? WHERE regulation_id = ? AND status = 'Pending'",
        (ts, reg_id),
    )
    conn.commit()
    rows = conn.execute("SELECT * FROM regulation_fields WHERE regulation_id = ? ORDER BY id", (reg_id,)).fetchall()
    conn.close()
    return {"data": [_field_dict(r) for r in rows]}
